docs/proposals/201911_thanos_query_mem_limits.md

```diff
@@ -0,0 +1,299 @@
+
+The [1] will need to be addressed a separate piece of work, if at all. There is already an effort to improve index for example.
+
+The approach to [2], [3], and [4] is *to approximate* and not aim for 100% accuracy. As long as the calculated memory use is in ballpark of the actual and tracks it well, this should do as a limit. The main challenge is therefore find the place in code where to place calculations and then figure out what exactly we need to calculate and how. 
+
+The approach to [5] is instrumentation with pprof profile heap snapshots at the right spots and comparing these figures with our calculations.
+
+
+## Pseudo code
+
+For calculations of sizes see separate section further.
+
+**Limiter interface:**
+
+```golang
+type Limiter interface {
+    Add(n int64) error
+}
+```
+
+
+**Querier:**
+
+```golang
+// Querier: pkg/store/proxy.go
+func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+    // Let each query source add their mem usage to this total limiter
+    queryTotalLimiter := limit.NewLimiterAtomic(limit.QueryTotalLimit())
+
+    for _, st := range s.stores() {
+        seriesSet := startStreamSeriesSet(..., queryTotalLimiter)
+    }
+}
+
+// Query of each store: startStreamSeriesSet
+func startStreamSeriesSet(..., queryTotalLimiter) {
+    // Have local limiter for this source
+    localLimiter limit.Limiter = limit.NewLimiterNoLock(limit.QueryPipeLimit())
+
+    // Receive frames
+    for {
+        r, err := s.stream.Recv()
+
+        // get the size of the received frame
+        frameSize := getApproxSize(r)
+
+        // check with local limiter first
+        if err := localLimiter.Add(frameSize); err != nil {
+            return err
+        }
+
+        // now add to total
+        if err := queryTotalLimiter.Add(frameSize); err != nil {
+            return err
+        }
+    }
+}
+```
+
+**Store Gateway:**
+
+```golang
+// pkg/store/bucket.go
+func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_SeriesServer) (err error) {
+    // Let each query source add their mem usage to this total limiter
+    queryTotalLimiter := limit.NewLimiterAtomic(limit.QueryTotalLimit())
+
+    for _, bs := range s.blockSets {
+        for _, b := range blocks {
+            part, pstats, err := blockSeries(..., queryTotalLimiter) {
+            }
+        }
+    }
+}
+
+func blockSeries(..., queryTotalLimiter) {
+    // Main memory users are lebels and aggrChunk. But not all. Only showing main parts
+
+    // Have local limiter here.
+    localLimiter limit.Limiter = limit.NewLimiterNoLock(limit.QueryPipeLimit())
+    
+    ps, err := indexr.ExpandedPostings(matchers)
+    for _, id := range ps {
+        frameSize := getFrameSize()
+
+        // check with local limiter first
+        if err := localLimiter.Add(frameSize); err != nil {
+            return err
+        }
+
+        // now add to total
+        if err := queryTotalLimiter.Add(frameSize); err != nil {
```

Comment:
Postings are not the only source of memory consumption. I assume we will do similar for series.

In fact while having series references we can be smart as we know the number of the chunks we are going fetch and we can add this before fetch even (:

Reply:
It would be nice if we could do some magic one-liner calculation. Probably best to look at the real code in https://github.com/thanos-io/thanos/pull/1747/files ?